corehq/apps/user_importer/importer.py

```diff
@@ -582,35 +580,60 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
                     # Passing use_primary_db=True because of https://dimagi-dev.atlassian.net/browse/ICDS-465
                     user.get_django_user(use_primary_db=True).check_password(password)
 
-                for group in domain_info.group_memoizer.by_user_id(user.user_id):
+                was_group_change = False
+                user_curr_groups = domain_info.group_memoizer.by_user_id(user.user_id)
+                user_curr_group_names = []
+                removed_from_groups = []
+                user_added_to_groups = []
+                # Remove user from groups and track the changes
+                for group in user_curr_groups:
+                    user_curr_group_names.append(group.name)
                     if group.name not in group_names:
+                        removed_from_groups.append(group)
+                        was_group_change = True
                         group.remove_user(user)
-
+                # Add user to groups and track the changes
                 for group_name in group_names:
-                    domain_info.group_memoizer.by_name(group_name).add_user(user, save=False)
+                    if group_name not in user_curr_group_names:
+                        user_added_to_groups.append(domain_info.group_memoizer.by_name(group_name))
+                        was_group_change = True
+                        domain_info.group_memoizer.by_name(group_name).add_user(user, save=False)
+
+                try:
+                    for domain_info in domain_info_by_domain.values():
+                        domain_info.group_memoizer.save_all()
+                except BulkSaveError as e:
+                    _error_message = (
+                        "Oops! We were not able to save some of your group changes. "
+                        "Please make sure no one else is editing your groups "
+                        "and try again."
+                    )
+                    logging.exception((
+                        'BulkSaveError saving groups. '
+                        'User saw error message "%s". Errors: %s'
+                    ) % (_error_message, e.errors))
+                    ret['errors'].append(_error_message)
+                # Collect updated group memberships and add change messages for updated groups to logger
+                if was_group_change:
+                    updated_groups = []
+                    for group in user_curr_groups:
+                        if group not in removed_from_groups:
+                            updated_groups.append(group)
+                    updated_groups.extend(user_added_to_groups)
+                    commcare_user_importer.logger.add_info(
+                        UserChangeMessage.groups_info(updated_groups)
+                    )
+
+                commcare_user_importer.save_log()
```

Comment:
This seems quite complex. I think it could be simplified:
```python
old_group_ids = set()
for group in domain_info.group_memoizer.by_user_id(user.user_id):
    old_group_ids.add(group.get_id)
    if group.name not in group_names:
        group.remove_user(user)

new_groups = {}
for group_name in group_names:
    group = domain_info.group_memoizer.by_name(group_name)
    group.add_user(user, save=False)
    new_groups[group.get_id] = group

if set(new_groups) != old_group_ids:
    commcare_user_importer.logger.add_info(UserChangeMessage.groups_info(list(new_groups.values()))
```

I think it would also be good to put this in a function on the user importer class:
```
error = commcare_user_importer.update_user_groups(domain_info, group_names)
if error:
    ret['errors'].append(error)

Reply:
You're right, it is pretty complex, thanks for the suggestion on simplifying. Moving the logic to `CommCareUserImporter` also makes a lot of sense.